src/main/kotlin/org/radarbase/push/integration/garmin/user/firebase/CovidCollabFirestore.kt

```diff
@@ -0,0 +1,197 @@
+
+    @Throws(IOException::class)
+    fun getUser(key: String): FirebaseUser? {
+        return try {
+            cachedUsers[key]
+        } catch (ex: NullPointerException) {
+            logger.warn("The requested user was not found in cache. Creating a new one.")
+            createUser(key)
+        }
+    }
+
+    fun getDocumentReferenceByServiceId(serviceId: String): DocumentReference? {
+        val uuid = getUsers().find { it.serviceUserId == serviceId }?.id ?: return null
+        return garminCollection.document(uuid)
+    }
+
+    @Throws(IOException::class)
+    protected fun createUser(uuid: String): FirebaseUser? {
+        logger.debug("Creating user using uuid...")
+        val garminDocumentSnapshot = getDocument(uuid, garminCollection)
+        val userDocumentSnapshot = getDocument(uuid, userCollection)
+        return createUser(userDocumentSnapshot, garminDocumentSnapshot)
+    }
+
+    protected fun createUser(
+        userSnapshot: DocumentSnapshot, garminSnapshot: DocumentSnapshot
+    ): FirebaseUser? {
+        if (!garminSnapshot.contains(OAUTH_KEY)) {
+            logger.warn(
+                "The $OAUTH_KEY key for user {} in the garmin" +
+                    " document is not present. Skipping...", garminSnapshot.id
+            )
+            return null
+        }
+
+        // Get the fitbit document for the user which contains Auth Info
+        val authDetails: FirebaseGarminAuthDetails? = garminSnapshot
+            .toObject(FirebaseGarminAuthDetails::class.java)
+        // Get the user document for the user which contains User Details
+        val userDetails: FirebaseUserDetails? = userSnapshot
+            .toObject(FirebaseUserDetails::class.java)
+
+        logger.debug("Auth details: {}", authDetails)
+        logger.debug("User Details: {}", userDetails)
+
+        // if auth details are not available, skip this user.
+        if (authDetails?.oauth2Credentials == null) {
+            logger.warn(
+                "The auth details for user {} in the database are not valid. Skipping...",
+                garminSnapshot.id
+            )
+            return null
+        }
+
+        return FirebaseUser(
+            uuid = garminSnapshot.id,
+            userId = garminSnapshot.id,
+            firebaseUserDetails = userDetails ?: FirebaseUserDetails(),
+            garminAuthDetails = authDetails
+        )
+    }
+
+    @Synchronized
+    private fun updateUser(garminDocumentSnapshot: DocumentSnapshot) {
+        try {
+            val user: FirebaseUser? = createUser(
+                getDocument(garminDocumentSnapshot.id, userCollection), garminDocumentSnapshot
+            )
+            logger.debug("User to be updated: {}", user)
+            if (checkValidUser(user)) {
+                val user1: FirebaseUser? = cachedUsers.put(user?.id ?: return, user)
+                if (user1 == null) {
```

Comment:
Would it be better to rename `user1` to something else (maybe `existingUser`)?

Reply:
Yes sure, updated now with some other improvements.